enterprise/backend/src/metabase_enterprise/remote_sync/impl.clj

```diff
@@ -0,0 +1,174 @@
+   [java-time.api :as t]
+   [metabase-enterprise.remote-sync.models.remote-sync-task :as remote-sync.task]
+   [metabase-enterprise.remote-sync.source :as source]
+   [metabase-enterprise.remote-sync.source.ingestable :as source.ingestable]
+   [metabase-enterprise.remote-sync.source.protocol :as source.p]
+   [metabase-enterprise.serialization.core :as serialization]
+   [metabase.analytics.core :as analytics]
+   [metabase.collections.models.collection :as collection]
+   [metabase.models.serialization :as serdes]
+   [metabase.util.log :as log]
+   [toucan2.core :as t2]))
+
+(defn- affected-collections
+  "Get all collections that are descendants of those in the sync and the ones synced"
+  []
+  ;; Otherwise get all remote-synced collections
+  (t2/select-pks-vec :model/Collection :type "remote-synced"))
+
+(defn sync-objects!
+  "Setup the remote-sync-object table with the imported-entities"
+  [timestamp imported-entities]
+  (t2/delete! :model/RemoteSyncObject)
+  (let [inserts (->> imported-entities
+                     (mapcat (fn [[model entity-ids]]
+                               (t2/select [(keyword "model" model) :id] :entity_id [:in entity-ids])))
+                     (map (fn [{:keys [id] :as model}]
+                            {:model_type (name (t2/model model))
+                             :model_id id
+                             :status "synced"
+                             :status_changed_at timestamp})))]
+    (t2/insert! :model/RemoteSyncObject inserts)))
+
+(defn- clean-synced!
+  "Delete any remote sync content that was NOT part of the import"
+  [synced-collection-ids imported-entities]
+  (when (seq synced-collection-ids)
+    (doseq [model [:model/Collection
+                   :model/Card
+                   :model/Dashboard
+                   :model/NativeQuerySnippet
+                   :model/Timeline
+                   :model/Document]
+            :let [serdes-model (name model)
+                  entity-ids (get imported-entities serdes-model [])]]
+      (if (= model :model/Collection)
+        (t2/delete! :model/Collection
+                    :id [:in synced-collection-ids]
+                    ;; if we didn't sync any, then delete all collections in the remote sync
+                    :entity_id (if (seq entity-ids)
+                                 [:not-in entity-ids]
+                                 :entity_id))
+        (t2/delete! model
+                    :collection_id [:in synced-collection-ids]
+                    :entity_id (if (seq entity-ids)
+                                 [:not-in entity-ids]
+                                 :entity_id))))))
+
+(defn- handle-import-exception
+  [e source-ingestable]
+  (if (:cancelled? (ex-data e))
+    (log/info "Import from git repository was cancelled")
+    (do
+      (log/errorf e "Failed to reload from git repository: %s" (ex-message e))
+      (analytics/inc! :metabase-remote-sync/imports-failed)
+      (let [error-msg (cond
```

Comment:
I think this is the same error handling that we have in the `/branches` API let's pull this into a separate function for reuse.